### PR TITLE
Enforcing your architecture with ArchUnit

### PR DESCRIPTION
**Pull request by Maiklins**
Please read through the [writing guide](https://reflectoring.io/contribute/writing-guide/) before submitting a pull request and go through this checklist (at least for your first article).

**Feel free to raise the pull request even if the checklist hasn't been worked through, yet**, to take advantage of the online preview that is generated for each pull request (see "deploy/netlify" under "Show all checks" in your pull request). But please only flag it to be reviewed by your editor after going through the checklist.

## Review
- [ ] you have reviewed the text a day after finishing it (you'd be surprised how many errors you can avoid by just getting some distance to what you've written)

## Content
- [ ] the article explains *why* and not only *how*
- [ ] (half-)sentences that express a meaningful idea that is understandable by itself own are **bold** to allow quick scanning of the text
- [ ] the text has been checked with [Grammarly](https://grammarly.com) (the free tier is enough)

## Text
- [ ] the text is conversational ("we" instead of "you", easy language)
- [ ] there are no "walls of text" (short sentences, short paragraphs with a single idea)
- [ ] the text is inclusive ("the developers" and "they" instead of "the developer" and "he/she")
- [ ] the text uses active voice instead of passive voice
- [ ] names are spelled consistently throughout the article (correct uppercase / lowercase)
- [ ] file names and variables are highlighted as `code`

## Structure
- [ ] the article has a short intro before the first headline
- [ ] by reading the table of contents (TOC) alone, I get a sense of the "story" of the article (check the TOC in the preview after raising the PR or by starting the blog locally on your machine)
- [ ] if using sub-headlines, there's at least two sub-headlines below a main-headline
- [ ] headlines are consistent throughout the article (same level of abstraction, same style - if one headline starts with a verb in imperative form, the other headlines on the same level should probably, too)

## Code Examples
- [ ] code examples are introduced with a ":" and explained below the code example
- [ ] code examples are formatted to avoid horizontal scrolling (manual line breaks and 2-spaces indentation)

### CODE REVIEW

File: content/blog/2023/2023-06-25-archunit.md [hunk 1]
```diff
@@ -0,0 +1,598 @@
...
+{{% image alt="Layer dependency" src="images/posts/enforce-architecture-with-arch-unit/arch-unit-layers.png" %}}
+(1) Shows the architecture without and (2) with a service layer. To keep the architecture clean, the API layer
+should only access the service layer and the service layer only access the domain layer. We should avoid direct 
+access from the API to the domain layer.
+
+## Basic ArchUnit Example
+Let's look at how we can use ArchUnit to create a test for the above example. 
+For that, we'll create a project with the following structure: 
+
+{{% image alt="Layer dependency" src="images/posts/enforce-architecture-with-arch-unit/example-project-structure.png" %}}
```
Comment (arkuksin) [inline content/blog/2023/2023-06-25-archunit.md:119]: For better understanding in this example, it is advisable to place the model classes within the layer packages. This approach would provide the reader with a clearer depiction of the layers.

File: content/blog/2023/2023-06-25-archunit.md [hunk 2]
```diff
@@ -0,0 +1,598 @@
...
+
+When we decide to use the latter way of instantiating our object, we'll probably remember the reason for a while. However, after some
+time we might forget. Also, other developers who join the project might unintentionally deviate from the original choice. 
+
+With ArchUnit, we can add a test that will fail if the static factory method ```now``` is used without the parameter: 
+
+```java
+@Test
+public void alwayInjectClockToFaciliateTesting() { 
+  // ...
```
Comment (arkuksin) [inline content/blog/2023/2023-06-25-archunit.md:221]: I would expect the entire test in this place.

File: content/blog/2023/2023-06-25-archunit.md [hunk 3]
```diff
@@ -0,0 +1,598 @@
...
+This is a good example of how we can use ArchUnit to document the intended architecture in the form of unit tests. Of course,
+we can change things when we see the need for it. There might be good reasons to deviate from a certain pattern. However,
+using tests as a documentation will remind us to think about why we want to deviate.
+
+## ArchUnit Examples 
+<a name="examples"></a>
+Most examples of how to use ArchUnit describe checks on dependencies between classes and packages.
+That's however not the only use case. Let's have a look at three other types of checks that we can create.
+
+### @deprecated
```
Comment (arkuksin) [inline content/blog/2023/2023-06-25-archunit.md:237]: Please modify the code to eliminate the need for horizontal scrolling by the reader in all code examples.

File: content/blog/2023/2023-06-25-archunit.md [hunk 4]
```diff
@@ -0,0 +1,598 @@
...
+
+@Deprecated
+public class Dep {
+}
+```
+With this test, we can check if we still use any deprecated methods. This check can be very useful in refactoring
+projects where we want to upgrade the version of libraries.
+
+### BigDecimal 
+Another nice use case is to **prevent the use of a specific constructor of a class**. Why would we won't to do this?
```
Comment (arkuksin) [inline content/blog/2023/2023-06-25-archunit.md:258]:
```suggestion
Another nice use case is to **prevent the use of a specific constructor of a class**. Why would we want to do this?
```

File: content/blog/2023/2023-06-25-archunit.md [hunk 5]
```diff
@@ -0,0 +1,598 @@
...
+public void do_not_call_constructor() {
+  JavaClasses importedClasses = new ClassFileImporter().importPackages("io.reflectoring.archunit");
+  ArchRule rule = noClasses().should().callConstructor(BigDecimal.class, double.class);
+  rule.check(importedClasses);
+}
+```
+    
+This test will fail if we call the BigDecimal constructor that accepts a double value as a parameter:
+```java
+public void this_method_calls_the_wrong_BigDecimal_constructor() {
```
Comment (arkuksin) [inline content/blog/2023/2023-06-25-archunit.md:279]: Is there a reason why you use the snake_case and not the camelCase

**Review by arkuksin — CHANGES_REQUESTED** [review]

File: content/blog/2023/2023-06-25-archunit.md (hunk 5 above)
Comment (Maiklins) [reply content/blog/2023/2023-06-25-archunit.md:279]: No, I have updated all method names.

File: content/blog/2023/2023-06-25-archunit.md (hunk 2 above)
Comment (Maiklins) [reply content/blog/2023/2023-06-25-archunit.md:221]: Added!

File: content/blog/2023/2023-06-25-archunit.md (hunk 1 above)
Comment (Maiklins) [reply content/blog/2023/2023-06-25-archunit.md:119]: I have updated the code and image.

File: content/blog/2023/2023-06-25-archunit.md [hunk 6]
```diff
@@ -0,0 +1,598 @@
...
+
+```java
+public record Employee(long id, String name, boolean active) { }
+```
+
+```java
+public class EmployeeController {
+  @GET()
+  @Path("/employees")
+  public EmployeeResponse getEmployee() {
```
[the remark this answers is not in the document]
Comment (Maiklins) [reply content/blog/2023/2023-06-25-archunit.md:58]: Indeed, good catch!

**Pull request merged**